pkg/ebpf/tracer.go

```diff
@@ -408,17 +408,36 @@ func (m *FlowFetcher) ReadRingBuf() (ringbuf.Record, error) {
 // This way we avoid missing packets that could be updated on the
 // ebpf side while we process/aggregate them here
 // Changing this method invocation by BatchLookupAndDelete could improve performance
-// TODO: detect whether BatchLookupAndDelete is supported (Kernel>=5.6) and use it selectively
-// Supported Lookup/Delete operations by kernel: https://github.com/iovisor/bcc/blob/master/docs/kernel-versions.md
 // Race conditions here causes that some flows are lost in high-load scenarios
 func (m *FlowFetcher) LookupAndDeleteMap(c prometheus.Counter) map[BpfFlowId][]BpfFlowMetrics {
 	flowMap := m.objects.AggregatedFlows
 
-	iterator := flowMap.Iterate()
 	var flows = make(map[BpfFlowId][]BpfFlowMetrics, m.cacheMaxSize)
+	var metrics = make([]BpfFlowMetrics, m.cacheMaxSize*ebpf.MustPossibleCPU())
 	var id BpfFlowId
-	var metrics []BpfFlowMetrics
+	var ids = make([]BpfFlowId, m.cacheMaxSize)
+	var cursor = ebpf.BatchCursor{}
+
+	for {
+		count, err := flowMap.BatchLookupAndDelete(&cursor, ids, metrics, nil)
+		if err == nil || errors.Is(err, ebpf.ErrKeyNotExist) {
+			for i, id := range ids[:count] {
+				for j := 0; j < ebpf.MustPossibleCPU(); j++ {
+					flows[id] = append(flows[id], metrics[i*ebpf.MustPossibleCPU()+j])
+				}
+			}
```

Comment:
Can you explain this, I'm not sure to understand?

Reply:
we are now getting percpu metrics so we need to combine all metrics for each CPU together and assign them to the right flow id